### ISSUE TITLE
bug positioning of webview in explorer

### ISSUE THREAD

**Issue by zqjimlove** [microsoft/vscode#110450]
- VSCode Version: 1.51
- OS Version:macOS 10.6.7

Steps to Reproduce:

My extension have 4 view. just the second view is webview.
1. expand the second view and the third view.
2. expand and shrink the first view, the second view will be broken

![屏幕录制2020-11-12 下午2 20 25](https://user-images.githubusercontent.com/3240259/98903538-c8886680-24f2-11eb-9594-0f9df64edcee.gif)

Does this issue occur when all extensions are disabled?: Yes/No

**Comment by tamuratak** [comment]
Version: 1.57.0-insider
Commit: 54d33cf14cf6c0e86880b5b74a5a3628de42bce9
Date: 2021-05-10T05:12:57.526Z
Electron: 12.0.4
Chrome: 89.0.4389.114
Node.js: 14.16.0
V8: 8.9.255.24-electron.0
OS: Darwin x64 18.7.0

I still see this issue.

1. Install https://marketplace.visualstudio.com/items?itemName=James-Yu.latex-workshop
2. Open a LaTeX file. You don't have to install TeX distributions.
3. Click the TeX button in the left side bar.
4. Expand commands, snipet view, and structure.


![スクリーンショット 2021-05-10 15 14 58](https://user-images.githubusercontent.com/10665499/117613923-8ccd1180-b1a2-11eb-9ee7-613ea14e210f.png)

**Comment by rzhao271** [comment]
@tamuratak I'm unable to reproduce your issue on the latest Insiders on Windows. Is it still showing up for you?

**Comment by tamuratak** [comment]
Yes, I still see the issue.

Version: 1.57.0-insider
Commit: a62df5d6c53e844bfdc50694f943c3f34b6ff50e
Date: 2021-06-03T17:36:36.265Z
Electron: 12.0.9
Chrome: 89.0.4389.128
Node.js: 14.16.0
V8: 8.9.255.25-electron.0
OS: Darwin x64 18.7.0

**Comment by tamuratak** [comment]
I also see the issue on Windows.

I have added the following step to reproducing steps.

> Move the snippet view in the middle between the structure view and the commands view.

Version: 1.57.0-insider (user setup)
Commit: a62df5d6c53e844bfdc50694f943c3f34b6ff50e
Date: 2021-06-03T17:38:02.890Z
Electron: 12.0.9
Chrome: 89.0.4389.128
Node.js: 14.16.0
V8: 8.9.255.25-electron.0
OS: Windows_NT x64 10.0.19042

**Comment by HuyQLuong** [comment]
Is there any improvement for this ? I also suffer the issue with webview in sidebar while I am trying to expand and collapse the view

**Comment by boltex** [comment]
Not for reproducing the bug: May happen only if a section **was moved**, and then, trying to open/close a section in the sidebar produces the bug. Meaning that if no section was moved, the bug may not appear.

Hope this helps isolating the bug.

**Comment by gbennardo** [comment]
Same problem here! 

@zqjimlove  Did you find a solution or a workaround to overcome the problem? 
@mjbvz Is your fix already available for the following version?

Thanks in advance

Versione: 1.58.2
Commit:  c3f126316369cd610563c75b1b1725e0679adfb3
Data: 2021-07-14T22:10:15.214Z
Electron: 12.0.13
Chrome: 89.0.4389.128
Node.js: 14.16.0
V8: 8.9.255.25-electron.0
OS: Windows or Ubuntu

**Comment by HuihuiWu-Microsoft** [comment]
I encounters the same issue. This issue happens when at least one view expanded below current webview section. I also reproduced this issue using the vscode sample 'webview-view-sample'. @tamuratak Do we have any plan to fix this?
![webview position](https://user-images.githubusercontent.com/73154171/133868451-7e86abf3-6dcf-412c-bf0a-c517fdb080ae.gif)

**Comment by a-stewart** [comment]
I posted this on the duplicate, copying here:

Hi,

I have noticed the same issue:

As far as I can see, is caused by the fact the webview is a child or the main element, not of the viewpane element.
When the view pane position is updated `layoutWebviewOverElement()` in `dynamicWebviewEditorOverlay.ts` to move the webview to the same position as the view pane.

However, the view pane is now animated, and the `layoutWebviewOverElement` is called instantly, which means the webview positition is set to where the view pane starts the animation, not where it finishes the animation.

For a really basic hack, you can update the layoutWebviewOverElement method to something like this:

```ts
	public layoutWebviewOverElement(element: HTMLElement, dimension?: Dimension, animated = true) {
		if (!this.container || !this.container.parentElement) {
			return;
		}

		const frameRect = element.getBoundingClientRect();
		const containerRect = this.container.parentElement.getBoundingClientRect();
		const parentBorderTop = (containerRect.height - this.container.parentElement.clientHeight) / 2.0;
		const parentBorderLeft = (containerRect.width - this.container.parentElement.clientWidth) / 2.0;
		this.container.style.position = 'absolute';
		this.container.style.overflow = 'hidden';
		this.container.style.top = `${frameRect.top - containerRect.top - parentBorderTop}px`;
		this.container.style.left = `${frameRect.left - containerRect.left - parentBorderLeft}px`;
		this.container.style.width = `${dimension ? dimension.width : frameRect.width}px`;
		this.container.style.height = `${dimension ? dimension.height : frameRect.height}px`;

		// Hack, there is an animation that lasts about 200ms, update the webview position afterwards.
		if (animated) {
			setTimeout(() => {
				this.layoutWebviewOverElement(element, dimension, false);
			}, 200);
		}
	}
```

I think a better solution, though, may be to either anchor the webview to the view pane element so that it animates with it.
Or to give the webview the details of the animation so that it can also animate in the same way.

**Comment by a-stewart** [comment]
Hi @mjbvz - I have created a PR with a temporary fix - it's not pretty but will (from a user point of view) fix this until a better solution can be implemented.

I'm also happy to work on a proper fix for it, but think it is best to sync with you first and see if you have any opinions on the best approach. One option would be rather than having a floating iframe, to make the iframe a child of the view so it would inherit the position, rather than manually having to keep it positioned correctly. (This will probably have to wait until the new year though.)

**Comment by a-stewart** [comment]
@mjbvz - I was wondering if you had any ideas for how best to handle this?

The PR: #137506 is hacky but will fix this until a proper solution can be created.

I'm happy to help with a less hacky solution, either moving the iframe into a child of the view container, or animating the floating iframe, but would like to know which direction you would like to go in before committing time to it.

**Comment by HuyQLuong** [comment]
If this problem still have not solved completely,  just suggest developers who faced this issue can try to create 1 webview only in side panel and implements the expandable menu by your self

I have tried and it work perfectly although it take time !

**Comment by boltex** [comment]
@HuyQLuong How did you do that? Can you share an example?

**Comment by HuyQLuong** [comment]
@boltex  let check in this repo https://github.com/HuyQLuong/vscode-webview-extension-with-react

**Comment by mjbvz** [comment]
Fixed by #110450

**Comment by phschaad** [comment]
> Fixed by #110450

I assume this is a typo.. :-) Where was it fixed - out of curiosity?
(Self-referencing this issue)

**Comment by boltex** [comment]
@mjbvz What? #110450 ??

**Comment by a-stewart** [comment]
I think he meant to say it was fixed by #137506 - although that PR was only really meant to be a workaround, and doesn't fully solve the issue, so I'd be tempted to leave this issue open until we have a fully solution that respects the animation.

A quick summary of the main underlying issue and why it occurs only under specific conditions.

We have a ResizeObserver which checks when the webview parent changes size, and updates the overlay webview accordingly. During a normal reveal, this observer fires multiple times making the webview appear to move with the view below it.

However, it is possible, by a certain arrangement of panels, to have the view move, but not change size. In this case, the ResizeObserver never fires, which is what the patch in #137506 handles, forcing the position to be manually recalculated 200ms later after animations are complete. This does not give us a smooth animation, but does ensure the webview ends up in the right place at the end of the animation.

Unfortunately, there is no such thing as a native movement observer to fire when an element is moved.
We could create one by polling and checking the bounding box, but I am not sure how expensive that would be.
